### PR TITLE
rgw: Report empty endpoints as error instead of crashing

### PR DESCRIPTION
**Pull request by adamemerson**
Empty endpoints are a misconfiguration, but should be rpeorted rather than crashing.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_trim_mdlog.cc [hunk 1]
```diff
@@ -677,9 +677,49 @@ class MetaPeerTrimPollCR : public MetaTrimPollCR {
   {}
 };
 
+namespace {
+bool sanity_check_endpoints(const DoutPrefixProvider *dpp, rgw::sal::RadosStore* store) {
+  bool retval = true;
+  auto current = store->svc()->mdlog->get_period_history()->get_current();
+  const auto& period = current.get_period();
+  for (const auto& [_, zonegroup] : period.get_map().zonegroups) {
+    if (zonegroup.endpoints.empty()) {
```
Comment (cbodley) [inline src/rgw/rgw_trim_mdlog.cc:686]: i don't think mdlog trimming ever tries to use zonegroup endpoints. it's fine to log a warning anyway, but this case probably shouldn't be fatal

File: src/rgw/rgw_trim_mdlog.cc [hunk 2]
```diff
@@ -677,9 +677,49 @@ class MetaPeerTrimPollCR : public MetaTrimPollCR {
...
+}
+
 RGWCoroutine* create_meta_log_trim_cr(const DoutPrefixProvider *dpp, rgw::sal::RadosStore* store, RGWHTTPManager *http,
                                       int num_shards, utime_t interval)
 {
+  if (sanity_check_endpoints(dpp, store)) {
+    ldpp_dout(dpp, -1)
+      << __PRETTY_FUNCTION__ << ":" << __LINE__
+      << " ERROR: Cluster is is misconfigured! Refusing to trim." << dendl;
+      return nullptr;
```
Comment (cbodley) [inline src/rgw/rgw_trim_mdlog.cc:721]: `sanity_check_endpoints()` returns false when not sane, so this condition should be negated. same with the call below

File: src/rgw/rgw_trim_mdlog.cc (hunk 1 above)
Comment (adamemerson) [reply src/rgw/rgw_trim_mdlog.cc:686]: Fixed! Thank you.

File: src/rgw/rgw_trim_mdlog.cc (hunk 2 above)
Comment (adamemerson) [reply src/rgw/rgw_trim_mdlog.cc]: Well, that was embarrassing.

**Review by cbodley — APPROVED** [review]

**Comment by adamemerson** [comment]
@cbodley https://pulpito.ceph.com/aemerson-2022-01-20_18:59:04-rgw-wip-53941-distro-basic-smithi/

**Pull request merged**